src/mlpack/methods/ann/rbm/spike_slab_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,332 @@
+      poolSize(poolSize),
+      slabPenalty(slabPenalty),
+      radius(2 * radius)
+{
+  parameter.set_size(visibleSize * hiddenSize * poolSize +
+      visibleSize + hiddenSize);
+
+  visibleMean.set_size(visibleSize, 1);
+  spikeMean.set_size(hiddenSize, 1);
+  spikeSamples.set_size(hiddenSize, 1);
+  slabMean.set_size(poolSize, hiddenSize);
+};
+
+// Reset function
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::Reset()
+{
+  // Weight shape D * K * N
+  weight = arma::Cube<ElemType>(parameter.memptr(),
+      visibleSize, poolSize, hiddenSize,
+      false, false);
+
+  // spike bias shape N * 1
+  spikeBias = DataType(parameter.memptr() + weight.n_elem, hiddenSize, 1,
+      false, false);
+
+  // visible penalty 1 * 1 => D * D(when used)
+  visiblePenalty = DataType(parameter.memptr() + weight.n_elem +
+      spikeBias.n_elem, 1, 1, false, false);
+}
+
+template<typename DataType>
+typename SpikeSlabRBMPolicy<DataType>::ElemType
+SpikeSlabRBMPolicy<DataType>::FreeEnergy(DataType&& input)
+{
+  assert(input.n_rows == visibleSize);
+  assert(input.n_cols == 1);
+
+
+  ElemType freeEnergy = 0.5 * arma::as_scalar(visiblePenalty(0) * input.t() *
+      input);
+
+  freeEnergy -= 0.5 * hiddenSize * poolSize *
+      std::log((2.0 * M_PI) / slabPenalty);
+
+  for (size_t i = 0; i < hiddenSize; i++)
+  {
+    ElemType sum = 0;
+
+    for (size_t k = 0; k < poolSize; k++)
+    {
+      sum += arma::as_scalar(input.t() * weight.slice(i).col(k)) *
+          arma::as_scalar(input.t() * weight.slice(i).col(k)) /
+          (2.0 * slabPenalty);
+    }
+
+    freeEnergy -= SoftplusFunction::Fn(spikeBias(i) - sum);
+  }
+
+  return freeEnergy;
+}
+
+template<typename DataType>
+typename SpikeSlabRBMPolicy<DataType>::ElemType
+SpikeSlabRBMPolicy<DataType>::Evaluate(DataType& /*predictors*/,
+    size_t /*i*/)
+{
+  // Return 0 here since we don't have evaluate in case of persistence
+  return 0;
+}
+
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>
+    ::PositivePhase(DataType&& input, DataType&& gradient)
+{
+  assert(input.n_rows == visibleSize);
+  assert(input.n_cols == 1);
+  arma::Cube<ElemType> weightGrad = arma::Cube<ElemType>
+      (gradient.memptr(), visibleSize, poolSize, hiddenSize, false, false);
+
+  DataType spikeBiasGrad = DataType(gradient.memptr() + weightGrad.n_elem,
+      hiddenSize, 1, false, false);
+
+  DataType visiblePenaltyGrad = DataType(gradient.memptr() +
+      weightGrad.n_elem + spikeBiasGrad.n_elem, 1, 1, false, false);
+
+  SpikeMean(std::move(input), std::move(spikeMean));
+  SampleSpike(std::move(spikeMean), std::move(spikeSamples));
+  SlabMean(std::move(input), std::move(spikeSamples), std::move(slabMean));
+
+  // positive weight gradient
+  for (size_t i = 0 ; i < hiddenSize; i++)
+    weightGrad.slice(i) = input * slabMean.col(i).t() * spikeMean(i);
+
+  // positive hidden bias gradient
+  for (size_t i = 0; i < hiddenSize; i++)
+    spikeBiasGrad(i) = spikeMean(i);
```

Comment:
Do we need the for loop here, looks like we could just do ``spikeBiasGrad = spikeMean`` since both is already of size ``hiddenSize``; maybe I missed something?